### ISSUE TITLE
Instruct tag format accuracy

### ISSUE THREAD

**Issue by Azirine** [LostRuins/koboldcpp#1032]
There's an issue with models where end of turn tokens are expected in the instruct tag. Take Llama 3 for example:

User Tag: `<|eot_id|><|start_header_id|>user<|end_header_id|>\n\n`
Assistant Tag: `<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n`

This works most of the time, except for the very first message, where there should be no <|eot_id|> in the user tag. This <|eot_id|> is not supposed to be before the user tag, but rather after the assistant's message. This works after the first message because 'after the assistant's message' is the same as 'before the user tag', but it's not the case for the first message.

[Debug: Dump Raw Input Tokens, format: 6]
'<|begin_of_text|> (128000)', **'<|eot_id|> (128009)'**, '<|start_header_id|> (128006)', 'user (882)', '<|end_header_id|> (128007)', '\n\n (271)', 'hi (6151)', '<|eot_id|> (128009)', '<|start_header_id|> (128006)', 'assistant (78191)', '<|end_header_id|> (128007)', '\n\n (271)',

The bolded token should not be there.

**Comment by LostRuins** [comment]
Yes, unfortunately this is a side effect of having only one tag instead of a "tag start" and "tag end".

To avoid this, perhaps you can put the very first message either manually in the Memory, or using the "System Prompt" instead, which has no `<|eot_id|>` that comes before it.

**Comment by Azirine** [comment]
Can you split the tag into "tag start" and "tag end" so that it works properly?

**Comment by Azirine** [comment]
I found another inaccuracy, this time in the system prompt format:

[Debug: Dump Raw Input Tokens, format: 6]
'<|begin_of_text|> (128000)', '<|start_header_id|> (128006)', 'system (9125)', '<|end_header_id|> (128007)', '\n\n (271)', 'test (1985)', **'\n\n (271)'**, '<|eot_id|> (128009)', '<|start_header_id|> (128006)', 'user (882)', '<|end_header_id|> (128007)', '\n\n (271)', 'hi (6151)', '<|eot_id|> (128009)', '<|start_header_id|> (128006)', 'assistant (78191)', '<|end_header_id|> (128007)', '\n\n (271)',

The bolded token should not be there.

This is Llama's example, there is no \n\n after system prompt.

```
<|begin_of_text|><|start_header_id|>system<|end_header_id|>

You are a helpful AI assistant for travel tips and recommendations<|eot_id|><|start_header_id|>user<|end_header_id|>

What can you help me with?<|eot_id|><|start_header_id|>assistant<|end_header_id|>
```

**Comment by jabberjabberjabber** [comment]
I think this actually belongs in [kobold lite](https://github.com/LostRuins/lite.koboldai.net/) and not koboldcpp. You might have to change from

```
"user":"<|eot_id|><|start_header_id|>user<|end_header_id|>\n\n",
"assistant":"<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n",
"system":"<|start_header_id|>system<|end_header_id|>\n\n",
```
to something like this one adapted from [clipboard conqueror](https://github.com/aseichter2007/ClipboardConqueror):
```
"startTurn": "<|start_header_id|>",
"endTurn": "<|eot_id|>",
"roleGap": "\n\n",
"systemRole": "system",
"endRole": "<|end_header_id|>",
"userRole": "user",
"assistantRole": "assistant",
```

but that would make the configuration more difficult and probably necessitate figuring out and applying each new chat template to each new model release.

**Comment by LostRuins** [comment]
Hi, in v1.92, split tags are now supported, so you should have a much better result. Additionally, automatically set tags are available as well (KoboldCppAutomatic mode).

Please reopen if issues persist.
